src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+                        logger.info("Saving checkpoint for step {} at {}".format(step, ckpt_save_path))
+
+                    if step % self.train_steps == 0:
+                        step += 1
+                        break
+
+                    step += 1
+
+                train_loss = total_loss / num_batches
+
+                self._evaluate_during_training()
+
+                logger.info("Epoch {} Step {} Train Loss {:.4f} Train Accuracy {:.4f}".format(epoch, step.numpy() - 1, train_loss.numpy(), self.train_acc_metric.result()))
+                logger.info("Epoch {} Validation Loss {:.4f} Validation Accuracy {:.4f}".format(epoch, self.test_loss_metric.result(), self.test_acc_metric.result()))
+
+            if epoch != self.epochs:
+                self.train_acc_metric.reset_states()
+                self.test_acc_metric.reset_states()
+
+    @abstractmethod
+    def _distributed_test_step(self, dist_inputs):
+        """
+        Method that represents a custom test step in distributed mode
+        Args:
+          dist_inputs: the features batch of the test data
+        """
+        pass
+
+    @abstractmethod
+    def _distributed_train_step(self, dist_inputs):
+        """
+        Method that represents a custom training step in distributed mode.
+        Args:
+          dist_inputs: the features batch of the training data
+        """
+        pass
+
+    @abstractmethod
+    def evaluate(self):
+        """
+        Evaluate the model over the test dataset and print a report.
+        """
+        pass
+
+    def save_model(self, save_path):
+        """
+        Save the pretrained model and create a Tensorflow saved model.
+        Args:
+          save_path: directory path where the pretrained model and
+            Tensorflow saved model will be saved
+        """
+        logger.info("Saving model in {}".format(save_path))
+
+        path = os.path.join(save_path, "saved_model")
+
+        os.makedirs(path, exist_ok=True)
+        self.model.save_pretrained(save_path)
+        tf.saved_model.save(self.model, path)
+
+
+class TFTrainerForSequenceClassification(TFTrainer):
+    def __init__(self, train_examples=None, dev_examples=None, test_examples=None, **config):
+        model_config = config.pop("model_config", None)
+        data_processor_config = config.pop("data_processor_config", None)
+        super().__init__(**model_config)
+        data_type = data_processor_config.pop("data_type", None)
+
+        if data_type == "csv":
+            self.processor = DataProcessorForSequenceClassificationWithCSV(**data_processor_config)
+        elif data_type == "tfds":
+            self.processor = DataProcessorForSequenceClassificationWithTFDS(train_examples, dev_examples, test_examples, **data_processor_config)
+        else:
+            raise ValueError(
+                "Allowed data types are csv or tfds. Given: {}".format(data_type)
+            )
```

Comment:
Would be nice to (1) reuse the dataprocessing from `pipelines` and (2) allow for streaming python data as well (list).

Reply:
Yep, agree, but the DataProcessor part has changed a lot, you will see in the next push. Anyway, yes I will try to better integrate this with what contain `pipelines`.